From b92d1dc29d8e6b542a1bcb1fa1d33dfb25045a16 Mon Sep 17 00:00:00 2001
From: Petros Angelatos <petrosagg@gmail.com>
Date: Thu, 3 Nov 2016 17:25:12 +0000
Subject: [PATCH] pkg/ioutils: sync parent directory too

After renaming the file to the target path the parent directory needs to
be synced to make sure the rename hits the disk.

Signed-off-by: Petros Angelatos <petrosagg@gmail.com>
---
 pkg/ioutils/fswriters.go | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/pkg/ioutils/fswriters.go b/pkg/ioutils/fswriters.go
index a56c462..7018476 100644
--- a/pkg/ioutils/fswriters.go
+++ b/pkg/ioutils/fswriters.go
@@ -5,6 +5,7 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"syscall"
 )
 
 // NewAtomicFileWriter returns WriteCloser so that writing to it writes to a
@@ -65,6 +66,9 @@ func (w *atomicFileWriter) Close() (retErr error) {
 			os.Remove(w.f.Name())
 		}
 	}()
+	if err := w.f.Chmod(w.perm); err != nil {
+		return err
+	}
 	if err := w.f.Sync(); err != nil {
 		w.f.Close()
 		return err
@@ -72,11 +76,22 @@ func (w *atomicFileWriter) Close() (retErr error) {
 	if err := w.f.Close(); err != nil {
 		return err
 	}
-	if err := os.Chmod(w.f.Name(), w.perm); err != nil {
-		return err
-	}
 	if w.writeErr == nil {
-		return os.Rename(w.f.Name(), w.fn)
+		if err := os.Rename(w.f.Name(), w.fn); err != nil {
+			return err
+		}
+
+		parentDir, err := os.OpenFile(filepath.Dir(w.fn), syscall.O_DIRECTORY, 0)
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			parentDir.Close()
+		}()
+		if err := parentDir.Sync(); err != nil {
+			return err
+		}
 	}
 	return nil
 }
-- 
2.10.2

